src/Sparrow/TimeValue.cs

```diff
@@ -0,0 +1,321 @@
+            if (remainingSeconds > 60)
+            {
+                var minutes = remainingSeconds / 60;
+                Append(str, minutes, "minute");
+                remainingSeconds -= minutes * 60;
+            }
+
+            if (remainingSeconds > 0)
+                Append(str, remainingSeconds, "second");
+
+            return str.ToString();
+        }
+
+        private void AssertMonthIsZero()
+        {
+            if (Months != 0)
+                throw new ArgumentException("Must be zero", nameof(Months));
+        }
+
+        internal void AssertMonthOrSeconds()
+        {
+            if (Months == 0 || Seconds == 0) 
+                return;
+
+            if (this == MaxValue || this == MinValue)
+                return;
+
+            throw new NotSupportedException($"Either {nameof(Months)} or {nameof(Seconds)} can be set.");
+        }
+
+        private const int SecondsIn28Days = 28 * SecondsPerDay; // lower-bound of seconds in month
+        private const int SecondsIn31Days = 31 * SecondsPerDay; // upper-bound of seconds in month
+
+        public int Compare(TimeValue other)
+        {
+            if (IsSpecialCompare(ref this, ref other, out var result))
+                return result;
+
+            if (Seconds == other.Seconds)
+                return TrimCompareResult(Months - other.Months);
+
+            if (Months == other.Months) 
+                return TrimCompareResult(Seconds - other.Seconds);
+
+            var myBounds = GetBounds(this);
+            var otherBounds = GetBounds(other);
+
+            if (otherBounds.UpperBound < myBounds.LowerBound)
+                return 1;
+
+            if (otherBounds.LowerBound > myBounds.UpperBound)
+                return -1;
+            
+            throw new InvalidOperationException("We can't compare ");
+        }
+
+        private static (int UpperBound, int LowerBound) GetBounds(TimeValue time)
+        {
+            var myUpperBound = time.Months * SecondsIn31Days + time.Seconds;
+            var myLowerBound = time.Months * SecondsIn28Days + time.Seconds;
+            return (myUpperBound, myLowerBound);
+        }
+
+        private static bool IsSpecialCompare(ref TimeValue current, ref TimeValue other, out int result)
+        {
+            result = 0;
+            if (IsMax(ref current))
+            {
+                result = IsMax(ref other) ? 0 : 1;
+                return true;
+            }
+
+            if (IsMax(ref other))
+            {
+                result = IsMax(ref current) ? 0 : -1;
+                return true;
+            }
+
+            if (IsMin(ref current))
+            {
+                result = IsMin(ref other) ? 0 : -1;
+                return true;
+            }
+
+            if (IsMin(ref other))
+            {
+                result = IsMin(ref current) ? 0 : 1;
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool IsMax(ref TimeValue time)
```

Comment:
Why is this a ref?

Reply:
It is a `struct`, but I guess he isn't big enough to benefit from passing it like this.

Reply:
Yes, it is same size as the pointer, no need to do this